### ISSUE TITLE
Function arguments desctructuring

### ISSUE THREAD

**Issue by NekR** [Microsoft/TypeScript#4548]
This does not work. This is expected since I have ``"noImplicitAny": true``:

```ts
function test({ x, y }) {
  console.log(x, y);
}

test({
  x: 1,
  y: 2
});
```

``src\main.ts(15,17): error TS7005: Variable 'x' implicitly has an 'any' type.``
``src\main.ts(15,20): error TS7005: Variable 'y' implicitly has an 'any' type.``

However, ``function test({ x:number, y:number }) {`` does not works too, while ``function test({ x = 0, y = 0 }) {`` works fine, but I do not want default params here. Is there a solution for it?

**Comment by xLama** [comment]
Which version of TS?

**Comment by NekR** [comment]
Latest stable I believe. 1.5.3

**Comment by xLama** [comment]
In this example:

```
function test({ x, y }) {
  console.log(x, y);
}

test({
  x: 1,
  y: 2
});

```
It showed "src\main.ts(15,17): error TS7005: Variable 'x' implicitly has an 'any' type." in 1.5.3 but now it is fixed in 1.7.0

And I don´t know what you can do with this

```
function test({ x:number, y:number }) { } 
```

You can´t define types in destructuring.

**Comment by NekR** [comment]
How it might be fixed if I use "noImplicitAny"? It should warn as it does.

Right, I cannot define type in destructuring, this is a problem. You may
call it a bug, if you wish.

What I can do with that number properties? Use them as number variables, as
expected, right?

**Comment by xLama** [comment]
As you can´t define type in destructuring, It make no sense that compiler shows an error if you set noImplicitAny to true.

I don´t think that is a bug. ES6 sintax uses " : " to rename property and it interferes with TS sintax type. 
To uses type it would provide other way:
```
function test({ x:number:renamedX , y:number:renamedY}) { } // Very very ugly 
```
Maybe could be...:

```
function test({ x:renamedX => number , y:renamedY => number}) { } // Very ugly too.
```

So, you can´t, for the moment

Yeah, you can:

```
function test({ x, y }) {
  console.log( x + y );
}

test({
  x: 1,
  y: 2
});
```

**Comment by ahejlsberg** [comment]
You can't have type annotations within a destructuring object literal because ES6 already defines a different meaning for `:` in a destructuring pattern. You can however have an annotation for the entire destructuring parameter:

```typescript
function test({ x, y }: { x: number, y: number }) {
  console.log(x, y);
}

test({
  x: 1,
  y: 2
});
```

See also #3828.
